### PR TITLE
apply relabel and drop_label rules before forwarding rules in distributor

### PR DESCRIPTION
**Pull request by replay**
Breaks the relabel/drop_label rules out of the `PushWithCleanup()` method and moves them into a middleware which gets applied at an earlier stage in the distributor's request processing.

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -864,27 +907,10 @@ func (d *Distributor) PushWithCleanup(ctx context.Context, req *mimirpb.WriteReq
...
-
-		for _, labelName := range d.limits.DropLabels(userID) {
-			removeLabel(labelName, &ts.Labels)
-		}
-
 		if len(ts.Labels) == 0 {
 			continue
 		}
 
-		// We rely on sorted labels in different places:
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:880]: Why is sorting of labels moved to `prePushRelabelMiddleware`? Is it to make sure we forward correct (sorted) labels?

File: pkg/distributor/distributor.go [hunk 2]
```diff
@@ -688,7 +685,57 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
...
+			// 2) In validation code, when checking for duplicate label names. As duplicate label names are rejected
+			// later in the validation phase, we ignore them here.
+			// 3) Ingesters expect labels to be sorted in the Push request.
+			sortLabelsIfNeeded(ts.Labels)
+		}
+
+		if len(removeTsIndexes) > 0 {
+			req.Timeseries, _, _ = util.RemoveSliceIndexes(req.Timeseries, removeTsIndexes)
+		}
+
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:737]: This is an optimization, while this PR should focus on refactor in the first place. I'd recommend to move optimizations to different PR, with a benchmark.

File: pkg/distributor/distributor.go [hunk 3]
```diff
@@ -688,7 +685,57 @@ func (d *Distributor) prePushHaDedupeMiddleware(next push.Func) push.Func {
...
+				ts.Labels = mimirpb.FromLabelsToLabelAdapters(l)
+			}
+
+			for _, labelName := range d.limits.DropLabels(userID) {
+				removeLabel(labelName, &ts.Labels)
+			}
+
+			if len(ts.Labels) == 0 {
+				if len(removeTsIndexes) == 0 {
+					// If we have to add one index into this slice then there is a good chance that we'll have to add more.
```
Comment (pstibrany) [inline pkg/distributor/distributor.go:718]: How do we know that? Series are sent in batches, using shards in Prometheus. Series in the same shard may not be related at all, so we could be allocating unnecessarily big slice here.

File: pkg/util/slice_operations.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+		// Extend the current range by one, because this index connected to it.
+		if currentRange.to+1 == index {
+			currentRange.to = index
+			continue
+		}
+
+		// The current range is finished, add it to the ranges list.
+		*ranges = append(*ranges, currentRange)
+
```
Comment (pstibrany) [inline pkg/util/slice_operations.go:63]: Why do we build list of ranges first, and don't just do the removal at this point?

File: pkg/util/slice_operations.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+// RemoveSliceIndexes takes a slice of elements of any type and a sorted slice of indexes of elements in the former slice.
+// It removes the elements at the given indexes from the given slice, while minimizing the number of copy operations
+// that are required to do so by grouping consecutive indexes into ranges and removing the ranges in single operations.
+// The given number of indexes must be sorted in ascending order.
+// Indexes which are duplicate or out of the range of the given slice are ignored.
+// The returned values are:
+// - The resulting slice of elements after the elements at the given indexes have been removed.
+// - The number of elements that have been removed.
+// - The number of ranges of consecutive elements that have been removed.
```
Comment (pstibrany) [inline pkg/util/slice_operations.go:28]: Why is number of ranges interesting for the caller?

File: pkg/util/slice_operations.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"sync"
+)
+
+type idxRange struct {
+	from int
+	to   int
+}
+
+var idxRangesPool = sync.Pool{
```
Comment (pstibrany) [inline pkg/util/slice_operations.go:14]: Why do we need a pool for this? Pools are not cheap: they make garbage collection slower, their use requires additional interface conversions.

Comment (pstibrany) [reply pkg/util/slice_operations.go]: Pool removed in https://github.com/grafana/mimir/pull/2703/commits/12a98bf242b0da30acd287a17cc0fbc2a43b5f59.

File: pkg/util/slice_operations.go (hunk 4 above)
Comment (pstibrany) [reply pkg/util/slice_operations.go]: Implemented in https://github.com/grafana/mimir/pull/2703/commits/12a98bf242b0da30acd287a17cc0fbc2a43b5f59.

**Review by pstibrany — APPROVED** [review]
The change makes sense to me.

I think we will need to handle validation in a similar way, so that we validate series before forwarding.

File: pkg/distributor/distributor.go [hunk 7]
```diff
@@ -996,6 +1022,19 @@ func copyString(s string) string {
 	return string([]byte(s))
 }
 
+// getConditionalCleanup returns two functions, the first one wraps the given cleanup function, the second one is a
+// function which makes the first one not call the given cleanup function if it has been called at least once.
+func getConditionalCleanup(cleanup func()) (func(), func()) {
```
[the remark this answers is not in the document]
Comment (replay) [reply pkg/distributor/distributor.go:1027]: fair enough, I'll change it back

File: pkg/util/slice_operations.go (hunk 4 above)
Comment (replay) [reply pkg/util/slice_operations.go]: you're right, building the list of ranges makes no sense 🤦‍♂️

File: pkg/util/slice_operations.go (hunk 5 above)
Comment (replay) [reply pkg/util/slice_operations.go:13]: it's mainly just for testability and there is barely any additional overhead in returning it.

File: pkg/util/slice_operations.go (hunk 6 above)
Comment (replay) [reply pkg/util/slice_operations.go]: thx!

File: pkg/distributor/distributor.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (replay) [reply pkg/distributor/distributor.go]: fixed here https://github.com/grafana/mimir/pull/2703/commits/d99bb4adfec2d89f612f2e2adad455214e8f4dbd

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (replay) [reply pkg/distributor/distributor.go:880]: Yes, because other middlewares after this one or the final push method might rely on the labels being sorted. 
if a relabel rule gets applied on a labelset then the result should always be sorted, but there is no guarantee that a relabel rule gets applied on every series, so I figured it would make sense to sort here because that way we can be sure that after the relabel middleware all labelsets are guaranteed to be sorted.
Do you think we should move it?

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (replay) [reply pkg/distributor/distributor.go:745]: I guess it could be called an optimization, I figured it's better if each middleware guarantees that its output is "clean" to prevent unexpected side effects if middlewares get combined differently in the future. Passing on time series without any labels seemed not clean.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (replay) [reply pkg/distributor/distributor.go:726]: I assumed that many series produced by one exporter usually have similar or even equal labels and they get scraped together in a single scrape, if drop_label/relabel rules ended up dropping every single label in a series then it's likely that they would do the same with other series of the same exporter because they likely have the same or a similar labels.

Maybe allocating a slice with then capacity of `len(req.Timeseries)` is a bit much, but any smaller number would just be a guess about how much might possibly be enough. What do you think we should do? Not pre-allocate the slice at all, or pre-allocate some small factor of `len(req.Timeseries)`, like `len(req.Timeseries)/4` or something like that?

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:880]: I think we should move it to "validation" middleware at some point, but let's keep it here for now.

Comment (pstibrany) [reply pkg/distributor/distributor.go:880]: I was mostly curious if I understand the reasoning.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:745]: By "optimization" I meant using batch-removal, not original removal that we had before.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (pstibrany) [reply pkg/distributor/distributor.go:726]: We don't use label relabeling in distributors ourselves, so it's not going to affect us too much either way.

Labels here include `__name__` label too, so if number of labels is zero, it means that the whole series is being dropped. I would say that's rare occurence, so starting with empty or small (eg. 10 elements) slice would work just fine here.

File: pkg/util/slice_operations.go (hunk 5 above)
Comment (pstibrany) [reply pkg/util/slice_operations.go:13]: Personally I would drop it from public API, but it's no big deal.

File: pkg/distributor/distributor.go (hunk 2 above)
Comment (replay) [reply pkg/distributor/distributor.go:745]: ah, ok. that's a fair point.

File: pkg/distributor/distributor.go (hunk 3 above)
Comment (replay) [reply pkg/distributor/distributor.go:726]: ok, let's start empty then. 
any number like `10` is a pure guess in my opinion, in that case we might as well rely on Go to be smart enough to grow the slice in reasonable steps.

Comment (replay) [reply pkg/distributor/distributor.go]: updated ae4b9ad542e0f06e2f6958d148dd981e80f67366

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (replay) [reply pkg/distributor/distributor.go:880]: > I think we should move it to "validation" middleware at some point, but let's keep it here for now.

that makes sense to me once we have the validation middleware

File: pkg/util/slice_operations.go (hunk 5 above)
Comment (replay) [reply pkg/util/slice_operations.go:13]: how would you drop it from the public API while still making sure that the function is testable? Make it private and then wrap it in a public function which drops this return value?

Comment (pstibrany) [reply pkg/util/slice_operations.go:13]: > Make it private and then wrap it in a public function which drops this return value?

Yes.

Comment (replay) [reply pkg/util/slice_operations.go]: how about this db981d937781fabebcaaeb317b75f14c5d58d5c6
